cmd/akamai-purger/main.go

```diff
@@ -0,0 +1,157 @@
+	urls := ap.toPurge[:]
+	ap.toPurge = []string{}
+	ap.mu.Unlock()
+	if len(urls) == 0 {
+		return
+	}
+
+	if err := ap.client.Purge(urls); err != nil {
+		// Add the URLs back to the queue?
+		ap.mu.Lock()
+		ap.toPurge = append(urls, ap.toPurge...)
+		ap.mu.Unlock()
+		ap.log.Errf("Failed to purge %d URLs: %s", len(urls), err)
+	}
+}
+
+// maxQueueSize is used to reject Purge requests if the queue contains
+// >= the number of URLs to purge so that it can catch up.
+var maxQueueSize = 1000000
+
+func (ap *akamaiPurger) Purge(ctx context.Context, req *akamaipb.PurgeRequest) (*corepb.Empty, error) {
+	ap.mu.Lock()
+	defer ap.mu.Unlock()
+	if len(ap.toPurge) >= maxQueueSize {
+		return nil, errors.New("Akamai purge queue too large")
+	}
+	ap.toPurge = append(ap.toPurge, req.Urls...)
+	return &corepb.Empty{}, nil
+}
+
+func main() {
+	grpcAddr := flag.String("addr", "", "gRPC listen address override")
+	debugAddr := flag.String("debug-addr", "", "Debug server address override")
+	configFile := flag.String("config", "", "File path to the configuration file for this service")
+	flag.Parse()
+	if *configFile == "" {
+		flag.Usage()
+		os.Exit(1)
+	}
+
+	var c config
+	err := cmd.ReadConfigFile(*configFile, &c)
+	cmd.FailOnError(err, "Reading JSON config file into config structure")
+
+	if *grpcAddr != "" {
+		c.AkamaiPurger.GRPC.Address = *grpcAddr
+	}
+	if *debugAddr != "" {
+		c.AkamaiPurger.DebugAddr = *debugAddr
+	}
+
+	scope, logger := cmd.StatsAndLogging(c.Syslog, c.AkamaiPurger.DebugAddr)
+	defer logger.AuditPanic()
+	logger.Info(cmd.VersionString())
+
+	clk := cmd.Clock()
+
+	tlsConfig, err := c.AkamaiPurger.TLS.Load()
+	cmd.FailOnError(err, "tlsConfig config")
+
+	if c.AkamaiPurger.PurgeInterval.Duration == 0 {
+		cmd.Fail("PurgeInterval must be > 0")
+	}
+
+	ccu, err := akamai.NewCachePurgeClient(
+		c.AkamaiPurger.BaseURL,
+		c.AkamaiPurger.ClientToken,
+		c.AkamaiPurger.ClientSecret,
+		c.AkamaiPurger.AccessToken,
+		c.AkamaiPurger.V3Network,
+		c.AkamaiPurger.PurgeRetries,
+		c.AkamaiPurger.PurgeRetryBackoff.Duration,
+		logger,
+		scope,
+	)
+	cmd.FailOnError(err, "Failed to setup Akamai CCU client")
+
+	ap := akamaiPurger{
+		client: ccu,
+		log:    logger,
+	}
+
+	stopped := make(chan bool, 1)
+	ticker := time.NewTicker(c.AkamaiPurger.PurgeInterval.Duration)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				ap.purge()
+			}
+		}
+	}()
+
+	serverMetrics := bgrpc.NewServerMetrics(scope)
+	grpcSrv, l, err := bgrpc.NewServer(c.AkamaiPurger.GRPC, tlsConfig, serverMetrics, clk)
+	cmd.FailOnError(err, "Unable to setup Akamai purger gRPC server")
+	akamaipb.RegisterAkamaiPurgerServer(grpcSrv, &ap)
+
+	go cmd.CatchSignals(logger, grpcSrv.GracefulStop)
+
+	err = cmd.FilterShutdownErrors(grpcSrv.Serve(l))
+	cmd.FailOnError(err, "Akamai purger gRPC service failed")
+
+	ticker.Stop()
+	timer := time.NewTimer(time.Second * 15)
+	select {
+	case <-timer.C:
+		cmd.Fail("Timed out waiting for purger to finish working")
+	case <-stopped:
```

Comment:
Those are the docs for `time.Timer`, `time.Ticker` doesn't have the same behavior. Stopping the ticker simply stops the channel from being populated with new ticks, it doesn't close the channel. If we want to gracefully shutdown we need to use some combination of channels to signal we are actually done, otherwise we would sit ranging over the ticker channel forever. The timeout is there to prevent the binary sitting for an extended period of time after catching a signal trying to finish any remaining purges, if we actually want that behavior we'd just remove the timeout.

Reply:
Ah, good point about the wrong docs. So I see the need for the `stop` chan.

> The timeout is there to prevent the binary sitting for an extended period of time after catching a signal trying to finish any remaining purges

The call to `grpcSrv.GracefulStop`, plus blocking on the return from `grpcSrv.Serve(l)`, means that by the time we get here, there are no RPCs remaining inflight. So I think we don't this timeout (though I agree we do need the `stopped` chan).

Reply:
Because of the batching no RPCs doesn't necessarily mean no work. It's possible there would be no RPCs but URLs still waiting to be purged.

Reply:
Ohh good call. So shouldn't the sleep time match the purgeinterval?

Reply:
No, but I think there is a bit of another bug here. The current timeout would only actually be useful if we call `Stop` during a call to `purge`. If we shut the purger down between two `purge` calls it will just break the loop and we will exit with URLs still in the queue as `Stop` doesn't actually cause a final tick, if it's between two ticks the last tick will just never happen. I think the solution here is to add a fallback call to `purge` just before writing to the `stopped` channel, outside of the loop.

This will catch a scenario where we stop between ticks and still have things left to purge. The timeout in both the case where we are in the middle of a `purge` call and between two `purge` calls ends up just being for the time it takes to actually send all the purge requests for akamai, I just pulled 15s out of the air.

Reply:
That makes sense; want to work on that fix?